Modules/Filtering/Smoothing/include/itkFFTDiscreteGaussianImageFilter.hxx

```diff
@@ -0,0 +1,202 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkFFTDiscreteGaussianImageFilter_hxx
+#define itkFFTDiscreteGaussianImageFilter_hxx
+
+#include "itkFFTPadImageFilter.h"
+#include "itkForwardFFTImageFilter.h"
+#include "itkInverseFFTImageFilter.h"
+#include "itkMultiplyImageFilter.h"
+
+#include "itkGaussianImageSource.h"
+#include "itkFFTConvolutionImageFilter.h"
+
+#include "itkNeighborhoodOperatorImageFilter.h"
+#include "itkGaussianOperator.h"
+#include "itkImageRegionIterator.h"
+#include "itkProgressAccumulator.h"
+#include "itkImageAlgorithm.h"
+#include "itkMacro.h"
+
+namespace itk
+{
+
+template <typename TInputImage, typename TOutputImage>
+void
+FFTDiscreteGaussianImageFilter<TInputImage, TOutputImage>::GenerateInputRequestedRegion()
+{
+  // call the superclass' implementation of this method. this should
+  // copy the output requested region to the input requested region
+  ImageToImageFilter<TInputImage, TOutputImage>::GenerateInputRequestedRegion();
+
+  // Get pointer to input
+  typename Superclass::InputImagePointer inputPtr = const_cast<TInputImage *>(this->GetInput());
+  if (inputPtr.IsNull())
+  {
+    return;
+  }
+
+  // get a copy of the input requested region (should equal the output
+  // requested region)
+  typename TInputImage::RegionType inputRequestedRegion;
+  inputRequestedRegion = inputPtr->GetRequestedRegion();
+
+  // pad the input requested region by the operator radius
+  typename TInputImage::SizeType radius;
+  radius.Fill(0);
+  for (size_t dim = 0; dim < ImageDimension; ++dim)
+  {
+    radius[dim] = (this->GetMaximumKernelWidth() - 1) / 2;
+  }
+  inputRequestedRegion.PadByRadius(radius);
+
+  // crop the input requested region at the input's largest possible region
+  if (inputRequestedRegion.Crop(inputPtr->GetLargestPossibleRegion()))
+  {
+    inputPtr->SetRequestedRegion(inputRequestedRegion);
+    return;
+  }
+  else
+  {
+    // store what we tried to request (prior to trying to crop)
+    inputPtr->SetRequestedRegion(inputRequestedRegion);
+
+    // Couldn't crop the region (requested region is outside the largest
+    // possible region).  Throw an exception.
+    itkExceptionMacro(<< "Requested region " << inputRequestedRegion
+                      << " is (at least partially) outside the largest possible region "
+                      << inputPtr->GetLargestPossibleRegion());
+  }
+}
+
+template <typename TInputImage, typename TOutputImage>
+void
+FFTDiscreteGaussianImageFilter<TInputImage, TOutputImage>::SetMaximumError(const ArrayType arr)
+{
+  itkWarningMacro("FFTDiscreteGaussianImageFilter uses a fixed-width Gaussian kernel and ignores maximum error");
```

Comment:
@thewtex @dzenanz Before moving out of WIP I'd like to get your take on handling elements from `DiscreteGaussianImageFilter` in its subclass here.

`DiscreteGaussianImageFilter` defines two variables with setters/getters for constructing a kernel:
- `MaximumError` is the allowable error resulting from kernel truncation
- `MaximumKernelWidth` is the maximum diameter of the Gaussian kernel and overrides `MaximumError` if necessary
These values are used by `itk::GaussianOperator` to construct a Gaussian kernel operator.

`FFTDiscreteGaussianImageFilter`, however, needs to construct a Gaussian kernel _image_ rather than an operator in order to apply FFT operations. `itk::GaussianImageSource` only accounts for image metadata (including size) and does not take kernel error into consideration.

Are the following interface adjustments that I've made here reasonable? If not, do you have suggestions for what could be changed?
- `MaximumError` is ignored and a warning is issued if the user tries to set its value
- `MaximumKernelWidth` is redefined from a loose constraint that _could_ be the resulting kernel size, to a hard constraint that explicitly sets the kernel size.

One alternative would be to revert to using an `itk::GaussianOperator` and then manually transfer operator values to a kernel image, but we would lose benefits of `itk::GaussianImageSource` such as defining the kernel off-centered from a pixel grid.

Reply:
What you are doing is reasonable. Possibly upgrade those warnings into exceptions/errors. This is a new class, and if someone switches to using it they can remove (or comment out) setting those ignored parameters.

Reply:
Thanks for the reassurance @dzenanz . My thinking behind making these warnings rather than errors is to allow smoothing to succeed if the user registers a factory to override `DiscreteGaussianImageFilter` with `FFTDiscreteGaussianImageFilter`, enabling the use of accelerated hardware. We're still "pretty close" to the original interface, so the user should be warned where parameters are unused but operations should be allowed to proceed.

Reply:
For an FFT Gaussian filter, `MaximumError` and `MaximumKernelWidth` could be used to determine how much to enlarge the `InputRequestedRegion`.

Reply:
@thewtex I'm not sure I understand, we're currently using `MaximumKernelWidth` as the fixed size to enlarge the `InputRequestedRegion`. Is the suggestion to go back to using `MaximumError` with a `GaussianOperator` to determine the kernel size?

Reply:
@thewtex we are already using `MaximumKernelWidth` to enlarge `InputRequestedRegion`: [`1f27fe4` (#3121) R62-R66](https://github.com/InsightSoftwareConsortium/ITK/pull/3121/commits/1f27fe425ed148b63dc71adb1faf148b5b3069a2#diff-35b3faf28f0ff02d401f61ba337a9a8b123e4449c85a21544894d946346d55ddR62-R66).

Reply:
> MaximumKernelWidth is the maximum diameter of the Gaussian kernel and overrides MaximumError if necessary

We should keep this behavior to determine radius to enlarge the InputRequestRegion. A `GaussianOperator` can be used to compute `GetRadius` as in the base class.

Reply:
Got it, thanks for clarifying @thewtex . Will revise.